### ISSUE TITLE
Percussion glyphs are not set/drawn

### ISSUE THREAD

**Issue by PieterHartzer** [opensheetmusicdisplay/opensheetmusicdisplay#325]
When loading a MusicXML file containing percussion (drums for example) the glyphs are not drawn correct.  It appears to use the default _keys: ["g/5"]_ instead of _keys: ["g/5/x2"]_

Expected notes to be similar to:
![image](https://user-images.githubusercontent.com/1129284/44114665-c5db9680-a003-11e8-8ed4-13d5983a7c21.png)

Render notes as:
![image](https://user-images.githubusercontent.com/1129284/44114515-4e54ba4c-a003-11e8-8c4b-d8d4f596e81c.png)

I am happy to attempt a fix if someone could point me in the right direction.

**Comment by PieterHartzer** [comment]
I think I might be getting somewhere with the implementation for this.  It still needs a little tidying up before I can submit a PR. 

VexFlowConverter.ts has the following method in which I need the note, not just the pitch.
`public static pitch(pitch: Pitch, clef: ClefInstruction) { ... }`
 
![image](https://user-images.githubusercontent.com/1129284/44179122-3f084500-a0ed-11e8-9332-2b0ac1562284.png)

**Comment by sschmidTU** [comment]
@PieterHartzer Looks great, thanks for helping out on this! The OSMD team doesn't have a lot of drums expertise, so it's nice to get some outside help.
Which information about the note do you need for the pitch? We could easily give the pitch method a GraphicalNote, because right now we just give it note.pitch:
```
this.vfpitch = VexFlowConverter.pitch(note.Pitch, this.clef);
```
So, we could give it note instead of note.Pitch. In theory, we could select the attributes we need (note.pitch, note.otherAttributeWeNeed...), but then we might need to change it again in the future. Why not give this small GraphicalNote parameter entirely.

It would be nice if you could create the PR already so we can look at the code together. It doesn't have to be ready to be merged yet, we often work on WIP PRs for a while until they're ready.

Also, i'm not 100% sure what the end result should look like. If you could upload an XML that displays correctly in Musescore, or an XML with correct rendering, that would be helpful.

**Comment by PieterHartzer** [comment]
@sschmidTU Thanks for the tip.  I added a _VexFlowGraphicalNote_ parameter which made it easier to pass through.  The pitch was still needed for transposing (I believe).  If I am getting something wrong, by all means point it out, I don't know much about music. 

I will create the PR once I work out how to do it with with Github across forks and branches.

Here is a screenshot of how it renders (I made a more realistic groove for most of it rather than random notes as before).
![image](https://user-images.githubusercontent.com/1129284/44290411-80752d80-a270-11e8-8e04-b1e840df5615.png)

And here is MuseCore:

![image](https://user-images.githubusercontent.com/1129284/44290394-68051300-a270-11e8-9e06-fc976420778e.png)

It doesn't look at great as MuseCore, but that is something I will perhaps pick up with VexFlow.

**Comment by sschmidTU** [comment]
I've merged your PR #327 to a new feature branch, feature/NoteHeadTypes. Your code is already very good and helpful, thanks for the contribution, @PieterHartzer !
I haven't merged to develop directly, because i would like to make a few changes, and in this project we usually work on features in branches and try to only merge clean, complete features into develop. But there isn't much missing for that.

My proposed changes:
* remove pitch argument from VexFlowConverter.pitch()
it's already included in the GraphicalNote parameter
* add missing [MusicXML note head types](from https://usermanuals.musicxml.com/MusicXML/Content/ST-MusicXML-notehead-value.htm) (as you mentioned in the code)
* Refactor noteHead.filled handling, create a NoteHeadType enum for filled and unfilled types.
* Create a new NoteHeadTypes test file with not only drum set notes, but all MusicXML-supported note head types

I will work on these, shouldn't take long. Of course, you are encouraged to keep contributing. You can create another PR at any time, e.g. for the new feature/NoteHeadTypes branch.

**Comment by sschmidTU** [comment]
By the way, MuseScore is not necessarily rendering everything correctly. For example, for the Cowbell, it renders the triangle sideways, which looks pretty, but doesn't seem to be standard notation.

![image](https://user-images.githubusercontent.com/33069673/44342165-6747c900-a48a-11e8-9af0-980b7d8ced6f.png)
(Musescore Top, OSMD bottom)

<https://en.wikipedia.org/wiki/Percussion_notation>

Though i'll see what i can do about the placement of the triangle, should connect to the stem.

**Comment by PieterHartzer** [comment]
I am glad to help.  
Your proposed changes make perfect sense.  

> Use Cross without square tips for Pedal Hi-Hat (D4) like in Musescore

I am busy working on a fix for this as the problem is in VexFlow and logged as [#640](https://github.com/0xfe/vexflow/issues/640)


>add missing MusicXML note head types (as you mentioned in the code)

From what I remember I am not convinced that VexFlow support the other ones but I might be wrong.


>Though i'll see what i can do about the placement of the triangle, should connect to the stem.

I believe this is a VexFlow issue too.  When the stem is down, it is too long.  See [their tests](http://www.vexflow.com/tests/?module=Percussion)

**Comment by sschmidTU** [comment]
There are some more note head types that Vexflow does support, like circle-x and a normal x. It seems Vexflow supports the following in the first line:
![image](https://user-images.githubusercontent.com/33069673/44346132-85ff8d00-a495-11e8-8ffe-f68050dc4e94.png)

Vexflow has a useful [visual regression test](http://www.vexflow.com/tests/) that shows some of these:
![image](https://user-images.githubusercontent.com/33069673/44346229-bc3d0c80-a495-11e8-991c-fa1d7963beec.png)

@PieterHartzer The above shows that it should be possible to use the cross without square tips, right?
Just use the right glyph code, i assume.

See the [table of glyphs Vexflow supports](https://user-images.githubusercontent.com/33069673/44346390-1c33b300-a496-11e8-9387-90e057e863fa.png),
which you can generate (with a white background) by executing npm install and grunt on the vexflow repository and opening glyphs.html. Unfortunately this doesn't seem to be available online.

I'm not sure we can use all of these for note heads, i'm also not sure you can use an undefined glyph head to show no note head "none", i'll test all of that.

About the stem and note head placement being a Vexflow problem, i thought as much, but it's good to know the specific Vexflow issue.

**Comment by PieterHartzer** [comment]
@sschmidTU  That is a lot more than what I could get out of VexFlow (inexperience perhaps).  I found the page with all the glyps which is super helpful.  I am very interested to know how you managed to get the first line all to display as is.  I could not get the X to be different without altering the VexFlow code which I haven't submitted yet (and might not need to if it works as it)

I haven't seen a ticket for the head/stem placement but noticed it in their regression tests which lead me to believe it's not a OSMD issue.

It's looking very good from the images! I cannot wait to see how you did it in the PR/branch :)

**Comment by sschmidTU** [comment]
@PieterHartzer sorry to disappoint, but the first image in the previous comment is from Musescore. I should have stated that.However, i'm pretty confident we can get most of that image implemented with the Vexflow glyphs.

**Comment by sschmidTU** [comment]
Closing the issue was not intentional. Apparently there is a keyboard shortcut to Close and comment.

**Comment by PieterHartzer** [comment]
@sschmidTU That makes me feel less incompetent :)  I will carry on with the changes in Vexflow and also see if I can get to the bottom of the stem issue, then raise a ticket for that too.  It would be great if we could get this rendering better.

**Comment by sschmidTU** [comment]
@PieterHartzer heh, see my edited comment above. I'm also looking forward to these improvements!

**Comment by sschmidTU** [comment]
I refactored the code in 565a0ef9b51316e50bcb62f5722e6c3f80e834fe, mainly to not use enum string values, also because "circle-x" is not a valid enum member name.
I'm not done yet, only wanted to commit my results for the day.

I combined the drum test samples into a drumset and a note head shape test file. The former is now the starting sample in the demo, that will be reset again when merging.

@PieterHartzer what did you want to do with the following lines? The if clause is never called, pitch is always defined. At least since my last commit. I guess you wanted to use the original pitch of the note instead of the transposed one for drum notes. But when we transpose, i think we also change the source note. For now, i removed these lines, because they are never called (i checked with a debugger). Also, the drum notes are (still) placed correctly.
```
        // If we are not setting the pitch (transpose) then use the note's
        if (pitch === undefined) {
            pitch = note.sourceNote.Pitch;
        }
```

By the way, i will merge develop into this branch tomorrow. There were 26 commits added, mainly a big new feature, expressions (dynamics etc).

**Comment by sschmidTU** [comment]
So far i didn't find a way to add square and slash shapes, which are present in the [vexflow glyphs](https://user-images.githubusercontent.com/33069673/44346390-1c33b300-a496-11e8-9387-90e057e863fa.png) (see above). You can't simply use the glyph code.

[Vexflow/tables.js](https://github.com/0xfe/vexflow/blob/master/src/tables.js) is interesting, unfortunately it only explicitly defines diamond, triangle and Cross/X glyphs (search for "diamond"). We could contribute the slash and square shape to it in a similar manner. I'm not sure if we can use them otherwise.

**Comment by PieterHartzer** [comment]
With regards to the `pitch === undefined` block.  I did that because in VexFlowGraphicalSymbolFactory -> addGraphicalAccidental the comments state 

> Sets a pitch which will be used for rendering the given graphical note (not changing the original pitch of the note!!!).

It then passes the pitch (transposed I presume) through, which then ends up in the if statement that I put in.

The code is looking good so far! 👍 

As for [Vexflow/tables.js](https://github.com/0xfe/vexflow/blob/master/src/tables.js), that is where I am fixing the current X head issue.  I could add the slash and square heads once I fixed the visual regression tests, or should I implement using the glyph code instead?

Looking on the web, I cannot see anywhere where the hollow X (cross) is referenced.  The bold cross is actually used although I cannot seem to create it in MuseScore.  See [Online Drummer](https://www.onlinedrummer.com/drum-key/)

**Comment by sschmidTU** [comment]
@PieterHartzer if you can add the slash and square heads to Vexflow, that would be great! Not sure about heads or glyph codes. I guess it would fit to the note_glyphs like triangle, diamond and X.

About the `pitch === undefined` block, i'll look into it. Still, if the code is never called, it's not necessary.

About the hollow/unfilled X, it's the same as the other note heads in Musescore, if the note duration is under a half note, it's unfilled, just like a normal half note. If you drop my drumset or shapes sample into MuseScore, it will show it, just like in the earlier screenshots above.

**Comment by PieterHartzer** [comment]
@sschmidTU The PR is in for VexFlow, but there are PRs open from months ago, so I am not sure whether it'll be in any time soon.

As for the work with the rest of the heads, I think it's better if we carry that discussion on on the ticket itself as opposed to here as it's getting rather busy on here already. [Here is a link to the ticket](https://github.com/0xfe/vexflow/issues/642)

I am not sure whether you actually get half notes on the cymbals, but if it's a standard then let's do it.

**Comment by sschmidTU** [comment]
I wasn't considering half notes on the drums specifically, i simply implemented showing the note heads as they are given in XML. If it's a half (or whole) note, the glyph is unfilled, except for special glyphs like the circle-x that have only one appearance. MuseScore behaves like this, too.

If someone writes a half note cymbal into XML, it's their responsibility ;)
![image](https://user-images.githubusercontent.com/33069673/44362854-9da13a80-a4c2-11e8-9bab-4847cb0ca141.png)

Thanks for opening the Vexflow issue, that's great. It seems open issues tend to get less frequent attention from the Vexflow people than finished PRs that solve them, so it's good you're already working on it. Tell me if you encounter any roadblocks, maybe i can help out on it.

**Comment by sschmidTU** [comment]
@PieterHartzer thanks to your great initial implementation, note head shapes are ready now (imo), see PR #337. Thanks again for your help, and i'm looking forward to see what happens with your [further note head additions in Vexflow#642](https://github.com/0xfe/vexflow/issues/642)!

**Comment by PieterHartzer** [comment]
@sschmidTU That's great news.  Thank you so much for your help too.  I made a comment on the PR, nothing major.  I added square and rectangle note heads but need to create some visual regression tests for them before I create the PR.  The cross is merge into Vexflow and can be pulled into OSMD.

**Comment by sschmidTU** [comment]
I just added the slash note head to our PR #337.
I didn't find a comment from you there, i guess you meant the Vexflow issue. Thanks for the hint of how to do a slash note in Vexflow by the way. It's pretty odd that you have to set it in the duration string in Vexflow instead of like other note heads.

About pulling Vexflow into OSMD, right now we rely on the npm release of Vexflow, which is 1.2.85, a recent release. We could use the master branch of Vexflow instead or a certain commit, but we'd have to build it ourselves, because afaik Vexflow builds are only commited for releases and such.
We've already thought about things like building Vexflow ourselves, but for now we prefer to see more frequent official releases from them, we've already talked to them about it.

**Comment by PieterHartzer** [comment]
The comment is in [VexflowConverter.js](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/pull/337/files#diff-fe9bbaffdcd6ce5a19ce1b7f5061d0e4) regarding `NoteHeadShape.Square` that is not handled.  It's not in Vexflow yet and hopefully will be in before the next release.

The slash is done slighly different, agreed.  It does however make a little sense considering it's done the same way as rests.  That way each duration could have a separate glyph, but it becomes a messy piece of code seeing as each note head has to be implemented for each duration.  It's implemented in different ways, presumably by different developers or a long time apart.

Vexflow in TypeScript would be nice.  This is the first TypeScript project I've ever worked on and I really like it.  It'll be very time consuming and having a full-time job doesn't help.  More releases would be good too.  It'll help get fixes across to OSMD faster.

I want to get the square and rectangle heads in (just tests remaining) as well as a fix for the stem being in the wrong place soon.  I cheated a little with the new note heads.  To make the stem appear in the right place, I had to make the heads the exact same width as the default glyph for note heads for `duration: '8'` notes until I get the fix for the stem in.  `shift_right` appears to be ignored (for `Flow.keyProperties.note_glyph` heads at least).

**Comment by sschmidTU** [comment]
Yes, VexFlow in Typescript would be great. It's often not so easy to see what a variable is if it doesn't have a type, especially for functions and class variables. We are sometimes fantasizing about porting VexFlow to Typescript. But as you said, it would be a lot of work. For now, we have a definition file for VexFlow (vexflow.d.ts), so at least in OSMD it's mostly typed. There is also a [DefinitelyTyped definition file for Vexflow](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/vexflow).

**Comment by sschmidTU** [comment]
@PieterHartzer the note head shapes for percussion are merged into develop now (#337). If you want to use this issue to continue talking about square and rectangle heads, i can leave it open for now.

**Comment by PieterHartzer** [comment]
@sschmidTU That's great news, thank you.  The title of this ticket is misleading in regards to the extra note heads.  Also, seeing as Vexflow might take a while to get them implemented, let's close this ticket and create a new one for implementing all the missing MusicXML noteheads instead.  I can't imagine it to be high priority as it's been missing from Vexflow for a while.
